### ISSUE TITLE
AI/ML Model Development

### ISSUE THREAD

**Issue by LufunoMphagi** [COS301-SE-2026/Mobile-Budgeting-App#224]
<html>
<body>
<html><head></head><body><h1>Build and Train Spending Forecast AI Model</h1><h2>Objective</h2><p>Develop the machine learning component responsible for predicting a user's projected month-end spending based on historical and current transaction behaviour.</p><p>The model must accept a fixed set of spending-related features from the Flutter application and return a predicted month-end spending amount.</p><p>The initial implementation should use TensorFlow/Keras for training and export the final trained model to TensorFlow Lite (<code inline="">.tflite</code>) so that it can eventually run locally within the mobile application.</p><hr><h2>Scope</h2><p>This task covers:</p><ul><li><p>Defining the machine learning problem.</p></li><li><p>Preparing or generating training data.</p></li><li><p>Performing feature preparation and normalization.</p></li><li><p>Creating and training the TensorFlow model.</p></li><li><p>Evaluating model accuracy.</p></li><li><p>Exporting the trained model to TensorFlow Lite.</p></li><li><p>Documenting the exact model input/output contract.</p></li><li><p>Providing the trained <code inline="">.tflite</code> model to the Flutter integration developer.</p></li></ul><p>This task does <strong>not</strong> include Flutter UI development or database integration.</p><hr><h2>Model Goal</h2><p>The model must predict:</p><p><strong>Predicted month-end spending</strong></p><p>Example:</p><p>Input:</p><ul><li><p>Current spending: R2,950</p></li><li><p>Average daily spending: R268</p></li><li><p>Previous month spending: R6,700</p></li><li><p>Three-month average: R6,300</p></li><li><p>Number of transactions: 27</p></li><li><p>Days elapsed: 11</p></li><li><p>Days remaining: 20</p></li></ul><p>Output:</p><ul><li><p>Predicted month-end spending: R7,180</p></li></ul><hr><h2>Required Model Input Features</h2><p>The initial model must use the following inputs in this exact order:</p>
Index | Feature | Description
-- | -- | --
0 | currentSpending | Total expenses recorded so far in the current month
1 | averageDailySpending | Current spending divided by number of elapsed days
2 | previousMonthSpending | Total expense amount for the previous completed month
3 | threeMonthAverage | Average monthly expenses across the previous three completed months
4 | transactionCount | Number of expense transactions in the current month
5 | daysElapsed | Number of days elapsed in the current month
6 | daysRemaining | Number of days remaining in the current month

<p>The Flutter integration will depend on this exact ordering.</p><p>Any change to the feature order must be communicated before integration.</p><hr><h2>Required Output</h2><p>The model must return one numeric value:</p><pre><code class="language-text">predictedMonthEndSpending
</code></pre><p>Example:</p><pre><code class="language-text">7180.50
</code></pre><p>The output should represent a monetary amount in the same unit as the model input.</p><hr><h2>Training Dataset</h2><p>Because there may not currently be enough real application users to train a reliable model, an initial synthetic financial dataset may be generated.</p><p>The dataset should represent realistic monthly spending behaviour.</p><p>Possible fields:</p><pre><code class="language-text">current_spending
average_daily_spending
previous_month_spending
three_month_average
transaction_count
days_elapsed
days_remaining
target_month_end_spending
</code></pre><p>The dataset should contain enough variation to represent:</p><ul><li><p>Low spending users</p></li><li><p>Moderate spending users</p></li><li><p>High spending users</p></li><li><p>Increasing spending patterns</p></li><li><p>Decreasing spending patterns</p></li><li><p>Stable spending patterns</p></li><li><p>Users early in the month</p></li><li><p>Users late in the month</p></li><li><p>Different transaction frequencies</p></li><li><p>Months with unexpected spending spikes</p></li></ul><p>Target dataset size:</p><pre><code class="language-text">20,000–100,000 records
</code></pre><p>The exact number may be adjusted depending on training performance.</p><hr><h2>Suggested Project Structure</h2><p>Create an ML workspace similar to:</p><pre><code class="language-text">ml/
├── datasets/
│   └── spending_training_data.csv
│
├── models/
│   ├── spending_forecast.keras
│   └── spending_forecast.tflite
│
├── generate_dataset.py
├── train_spending_model.py
├── evaluate_model.py
└── README.md
</code></pre><hr><h2>Dataset Generation</h2><p>Implement:</p><pre><code class="language-text">generate_dataset.py
</code></pre><p>Responsibilities:</p><ul><li><p>Generate realistic spending scenarios.</p></li><li><p>Generate target month-end spending values.</p></li><li><p>Prevent invalid values such as negative days or impossible date combinations.</p></li><li><p>Add realistic noise so that the model does not simply learn a fixed formula.</p></li><li><p>Save the generated dataset to CSV.</p></li></ul><p>The generated data should not contain personally identifiable information.</p><hr><h2>Model Training</h2><p>Implement:</p><pre><code class="language-text">train_spending_model.py
</code></pre><p>Responsibilities:</p><ul><li><p>Load the training dataset.</p></li><li><p>Separate input features from target values.</p></li><li><p>Split data into training, validation, and test sets.</p></li><li><p>Normalize or scale numerical features where required.</p></li><li><p>Create the TensorFlow/Keras model.</p></li><li><p>Train the model.</p></li><li><p>Monitor validation performance.</p></li><li><p>Prevent excessive overfitting.</p></li><li><p>Save the final trained model.</p></li></ul><p>The model does not need to be unnecessarily complex.</p><p>A small dense neural network is sufficient for the first implementation.</p><hr><h2>Model Evaluation</h2><p>Implement:</p><pre><code class="language-text">evaluate_model.py
</code></pre><p>The evaluation should include appropriate regression metrics such as:</p><ul><li><p>Mean Absolute Error (MAE)</p></li><li><p>Mean Squared Error (MSE)</p></li><li><p>Root Mean Squared Error (RMSE)</p></li></ul><p>The developer should report the final evaluation results.</p><p>Example:</p><pre><code class="language-text">Test MAE: R320.50
Test RMSE: R445.20
</code></pre><p>The model should also be manually tested using several realistic input examples.</p><hr><h2>TensorFlow Lite Export</h2><p>Convert the final trained model into:</p><pre><code class="language-text">spending_forecast.tflite
</code></pre><p>The exported model must:</p><ul><li><p>Load successfully using TensorFlow Lite.</p></li><li><p>Accept exactly 7 numeric input features.</p></li><li><p>Return exactly 1 numeric output.</p></li><li><p>Produce predictions close to the original TensorFlow/Keras model.</p></li></ul><p>The <code inline="">.tflite</code> file must be committed or made available for integration into the Flutter application.</p><hr><h2>Model Contract Documentation</h2><p>Create documentation explaining:</p><h3>Input shape</h3><p>Example:</p><pre><code class="language-text">[1, 7]
</code></pre><h3>Input order</h3><pre><code class="language-text">[
  currentSpending,
  averageDailySpending,
  previousMonthSpending,
  threeMonthAverage,
  transactionCount,
  daysElapsed,
  daysRemaining
]
</code></pre><h3>Output shape</h3><p>Example:</p><pre><code class="language-text">[1, 1]
</code></pre><h3>Output meaning</h3><pre><code class="language-text">Predicted month-end spending amount
</code></pre><h3>Data normalization</h3><p>Document whether the Flutter application needs to normalize input before inference.</p><p>If normalization is required, provide the exact calculation or constants.</p><hr><h2>Privacy Requirements</h2><p>The initial model must not require uploading a user's raw transaction history to a remote AI service.</p><p>The model should operate using aggregated numeric features only.</p><p>No sensitive user details such as:</p><ul><li><p>Transaction descriptions</p></li><li><p>Merchant names</p></li><li><p>Account numbers</p></li><li><p>User names</p></li><li><p>Bank information</p></li></ul><p>should be required by the model.</p><hr><h2>Testing Requirements</h2><p>Tests or validation scripts should verify:</p><ul><li><p>Model accepts exactly 7 input features.</p></li><li><p>Model produces a valid numeric prediction.</p></li><li><p>Model output is not NaN or infinite.</p></li><li><p>TensorFlow Lite output is reasonably close to TensorFlow/Keras output.</p></li><li><p>Invalid or extreme test inputs do not cause crashes.</p></li></ul><hr><h2>Deliverables</h2><p>The completed task should provide:</p><ul><li><p>Synthetic training dataset or generation script.</p></li><li><p>TensorFlow training script.</p></li><li><p>Model evaluation script.</p></li><li><p>Trained Keras model.</p></li><li><p>Exported <code inline="">.tflite</code> model.</p></li><li><p>Model evaluation results.</p></li><li><p>Model input/output documentation.</p></li><li><p>Instructions for running the training pipeline.</p></li></ul><hr><h2>Acceptance Criteria</h2><p>This issue is complete when:</p><ul class="contains-task-list"><li class="task-list-item"><p><input type="checkbox" disabled=""> Training dataset has been created.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> TensorFlow model has been successfully trained.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model uses the agreed 7 input features.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model predicts month-end spending.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model performance has been evaluated.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <code inline="">.tflite</code> model has been generated.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> <code inline="">.tflite</code> model loads successfully.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> TensorFlow Lite predictions have been verified.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Input/output contract has been documented.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Normalization requirements have been documented.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Model artifact has been provided to the Flutter integration developer.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> Relevant code is covered by appropriate validation/tests.</p></li><li class="task-list-item"><p><input type="checkbox" disabled=""> README/instructions explain how to reproduce the training process.</p></li></ul><hr><h2>Suggested Branch</h2><pre><code class="language-text">feature/ai-spending-model
</code></pre><h2>Dependencies</h2><p>This task can be developed independently initially.</p><p>The final <code inline="">.tflite</code> model and model specification must be provided to the developer responsible for Flutter model integration.</p></body></html>
</body>
</html>
